node/src/actors/chain_manager/mod.rs

```diff
@@ -161,6 +161,138 @@ pub struct SyncTarget {
     superblock: CheckpointBeacon,
 }
 
+#[derive(Debug, Default)]
+struct ChainStateSnapshot {
+    // Previous chain_state, and the superblock index that will confirm the consensus
+    // When superblock with index n is consolidated by the ChainManager, the previous_chain_state
+    // with superblock index n should be irreversibly persisted into the storage
+    previous_chain_state: Option<(ChainState, u32)>,
+    // The ChainState at this superblock index is already persisted in the storage
+    highest_persisted_superblock: u32,
+}
+
+impl ChainStateSnapshot {
+    // Returns false if the snapshot did already exist
+    // Returns true if the snapshot did not already exist
+    // Panics if a different chain state was already saved for this super epoch
+    fn take(&mut self, superblock_index: u32, state: &ChainState) -> bool {
+        let chain_beacon = state
+            .chain_info
+            .as_ref()
+            .expect("ChainInfo is None")
+            .highest_block_checkpoint;
+
+        let superblock_beacon = state
+            .chain_info
+            .as_ref()
+            .expect("ChainInfo is None")
+            .highest_superblock_checkpoint;
```

Comment:
There are 2 functions that do the same (get_beacon(), and get_superblock_beacon)

Reply:
Yes, sadly this is not the same as `self.get_beacon()`, because we are getting the beacon of the new `state`, and `state.get_beacon()` does not exist.

Reply:
Done, created `ChainState::get_chain_beacon` and `ChainState::get_superblock_beacon`